Added vacation message to profile, show vacation icon on Overview & Game pages

dwvanstone opened the pull request:
Fixes #1378 .

It'll be a while before I get a test to run successfully, but when I do I'll let you know!

blackshadowshade commented:
Okay, a couple of procedural things:
- Because of the way Chaos runs database updates, the views need to be regenerated within the updates. You'll want to do the whole drop and create of both views.
- We always state that a database update is required (and which one) in the header text of a pull request, just as a reminder for Chaos that she needs to run it.

blackshadowshade commented:
I also note that you have use onPlayer as an implicit property of BMPlayer without actually defining it within BMPlayer. This should really be added as an explicit property with appropriate documentation of type and intent.

blackshadowshade commented on deploy/database/schema.player.sql:
```diff
@@ -37,6 +37,7 @@ CREATE TABLE player (
     fanatic_button_id   SMALLINT UNSIGNED,
     n_games_won         SMALLINT UNSIGNED DEFAULT 0,
     n_games_lost        SMALLINT UNSIGNED DEFAULT 0,
+    vacation_message	VARCHAR(255),
```
Nit: spacing.
Also, is 255 characters long enough for a vacation message?

blackshadowshade commented on deploy/database/updates/01378_vacation_message.sql:
```diff
@@ -0,0 +1 @@
+ALTER TABLE player ADD vacation_message VARCHAR(256);
```
This size doesn't match the size in schema.player.sql.

blackshadowshade commented on deploy/database/views.game.sql:
```diff
@@ -22,6 +22,7 @@ AS SELECT
     p.name_ingame AS player_name,
     p.autopass,
     p.fire_overshooting,
+    (length(p.vacation_message)>0) as on_vacation,
```
Nit: add single spaces around the '>'

blackshadowshade commented on src/ui/gui.css:
```diff
@@ -453,6 +453,14 @@ table.activePlayerList th.idleness {
 
+.playerFlag {
+   font-weight: bold;
+	   font-style: italic;*/
```
What's going on here?

dwvanstone replied:
Just badness. Utter badness. Fixable, thankfully.

blackshadowshade commented on src/ui/js/Game.js:
```diff
@@ -2562,6 +2562,9 @@ Game.buttonImageDisplay = function(player) {
   var playerName = $('<div>', {
     'html': $('<b>', { 'text': 'Player: ', }),
   });
+  if (Api.game[player].onVacation>0) {
```
Nit: add single spacing around the '>'

blackshadowshade commented on src/ui/js/Game.js:
```diff
@@ -2892,7 +2895,13 @@ Game.playerOpponentHeaderRow = function(label, field) {
   var playerInfo = $('<th>', { 'text': prefix, });
   var opponentInfo = $('<th>', { 'text': prefix, });
   if (field == 'playerName') {
+    if (Api.game.player.onVacation>0) {
```
Nit: spacing

blackshadowshade commented on src/ui/js/Game.js:
```diff
@@ -2892,7 +2895,13 @@ Game.playerOpponentHeaderRow = function(label, field) {
+       playerInfo.append(Env.buildVacationImage());
+    }
     playerInfo.append(Env.buildProfileLink(Api.game.player[field]));
+    if (Api.game.opponent.onVacation>0) {
```
And again

blackshadowshade commented on src/ui/js/Overview.js:
```diff
@@ -389,11 +390,12 @@ Overview.addButtonCol = function(gameRow, buttonName) {
       'style': 'background-color: ' + playerColor,
-    }).append(Env.buildProfileLink(playerName)));
-};
+    }).append((playerOnVacation>0) ? Env.buildVacationImage() : '')
```
Also here, I'm thinking that you don't need the '>0'

blackshadowshade commented:
Okay, that's it for my first look through your code. I'll let Chaos actually test functionality.  :)

dwvanstone commented:
I wouldn't go ahead and test functionality - I haven't been able to get a clean test yet. I have to figure that out first. Also, I might as well clean up the bits you've found. (Thanks for looking at it!)

cgolubi1 commented:
> * We always state that a database update is required (and which one) in the header text of a pull request, just as a reminder for Chaos that she needs to run it.

So, i've forgotten what it is that i broke way back when that caused me to ask for this reminder, but i think we have database updates better organized now, and i don't actually need the text reminder in the pull --- as long as the SQL updates/ files are present and do the right thing, i should do the right thing during deployments too.

blackshadowshade commented:
Okay, good to know.

dwvanstone commented:
Okay - I believe I have fixed all of the minor nits. The closest Jenkins I have to passing is my latest:

http://jenkins.buttonweavers.com:8080/job/buttonmen-dwvanstone/36/

It fails one test:
responderTest::test_request_loadActiveGames
Real and dummy game lists should have matching structures

I don't know how to fix that one failure - it has to do with the dummy game list not being right. Maybe you can go ahead and test behavior even with this one failure?

blackshadowshade commented:
Right, try merging in upstream/master now and resolving the conflicts, and see if that fixes your problem.

dwvanstone commented:
Now I have to learn how to merge! Thanks for checking in the changes...

dwvanstone commented:
That took a while, but I have a clean test now:

http://jenkins.buttonweavers.com:8080/job/buttonmen-dwvanstone/45/

blackshadowshade commented:
14 iterations has to be approaching a record!  ;)

blackshadowshade commented:
So, the last thing left from my point of view is the question that we discussed in chat a while back, namely whether isOnVacation belongs in BMPlayer or not. I'm happy with it being there, and if so, it should be declared explicitly within BMPlayer, with a comment about its purpose and that it's only used within BMInterface.

dwvanstone commented:
Okee-doke, that should be fine. I'll work on that next.

blackshadowshade commented:
I've assigned this to @cgolubi1, since I think it's now ready for functionality testing.

dwvanstone commented:
Here's the next passing Jenkins:

http://jenkins.buttonweavers.com:8080/job/buttonmen-dwvanstone/46/

cgolubi1 commented on deploy/database/views.game.sql:
```diff
@@ -22,6 +22,7 @@ AS SELECT
     p.name_ingame AS player_name,
     p.autopass,
     p.fire_overshooting,
+    (length(p.vacation_message) > 0) as is_on_vacation,
```
Something i'll be looking for in testing - since there's no separate option for *whether* a player is on vacation other than the message itself, is it easy for a player to remove their vacation message and see that they've done so?  It'd be frustrating if you left some whitespace in the box and were still on vacation and couldn't figure out why.

dwvanstone replied:
White space gets stripped out, so it won't be a whitespace option. There'a a few subtle places to catch whether a user is on vacation: if they look at their own profile, the Vacation Message will appear only if they are on vacation. Also, a user can see their own vacation status when they're playing a game. I started thinking about putting vacation up on the top ("You are logged in as USER. You are on vacation") but that was starting to become difficult. I'm happy to entertain suggestions!

cgolubi1 commented on src/engine/BMPlayer.php:
```diff
@@ -462,6 +470,20 @@ protected function set__ownerObject($value) {
+
+    /**
     // utility methods
     /**
```
Seems like a stray `/**` got inserted here.

cgolubi1 commented:
Okay, so, how am i going to test this?
* Standard "database update vs database rebuild" test to make sure the database update file is correct
* Poke around in the UI for a bit and see what it looks like:
    * Any issue with the UI for updating vacation messages?
    * If i am on vacation and browse around the site, does anything look any different?
    * If i have games against an opponent who is on vacation, does everything look okay?  (What if i have games vs both on-vacation and not-on-vacation opponents in the overview?  What if someone who is on vacation creates an open game?  Do i like the look of the overview UI, the game UI, the profile UI, the forum UI?)
* See if i can break the changes to BMInterface:
    * Try setting some or all of the responderTest fake players to be on vacation.  Do all the API and UI tests still succeed?
* We should lock in the end-to-end behavior by actually setting one of the responderTest players to be on vacation in the committed tests, and having a UI test using that information.  Based on my testing in the previous step, i'll make a recommendation for what that should be.

cgolubi1 commented:
Okay, so there's definitely a toggle problem; it's not quite what i thought it was, it's worse --- whenever you save your preferences from the UI, if there's nothing in the vacation_message box (or just whitespace), the UI submits vacation_message='' (empty string).  The backend considers empty string to be different from NULL.  So as soon as you modify your preferences, you are on vacation forever.  So, that needs a fix.

cgolubi1 commented:
Out of testing time for today; more tomorrow.

dwvanstone commented:
Got the toggle problem - it's a PHP subtlety I'm learning. The string value ("0") was being passed to Javascript which says that "0" is TRUE, and I needed to send the boolean value (0) instead. Sloppy testing on my part, for which I apologize. (Toggling used to work some iterations ago...!)

I'll report when I get a working Jenkins. I accidentally did a "merge branch of master" which I'm hoping won't interfere. (I got all of the Zodiac images.)

dwvanstone commented:
http://jenkins.buttonweavers.com:8080/job/buttonmen-dwvanstone/48/ has passed.

dwvanstone commented:
Updated something small (to enfore the BMPlayer changes), assuming test will pass:

http://jenkins.buttonweavers.com:8080/job/buttonmen-dwvanstone/49/

cgolubi1 commented:
Makes sense.  I'll redo the database test due to the merge, but i don't expect any problems.

cgolubi1 commented:
I think it would be nice to have something on the profile page, and maybe also on the preferences page but definitely on the profile page, to draw your eye to the fact that the player is on vacation.  For instance, if the "Vacation Message" line appears, it could have an orange background.  Or you could put the airplane next to the player's name at the top, if you want to consistently use the airplane as a vacation identifier --- i'd be inclined to go with color on the profile page, where there currently isn't any so it'd be eye-catching, but i could see either way working.

cgolubi1 commented:
Anyway, confirmed that the toggling problem appears fixed.  Also confirmed that the vacation message box eats whitespace correctly, so if i put a bunch of spaces and returns in the box, i am still not on vacation.

cgolubi1 commented:
Another nice to have --- on the Open Games page, show the airplane icon where appropriate in the table of open games.

cgolubi1 commented:
So, to be a bit picky about the API: i don't like the fact that when you're not on vacation, the API reports `isOnVacation=0`, and when you are on vacation it reports `isOnVacation=true`.  I think the return value should be boolean (false or true) in both cases, not numeric sometimes and boolean sometimes.  This is minor, but this kind of thing is easier to fix at the outset than to make time to fix later.

cgolubi1 commented:
Ah.  So, i got a failure running the QUnit tests on Firefox on my mac, which is somewhat instructive about how HTML is brittle:
![ui_test_failure](https://cloud.githubusercontent.com/assets/2898207/13555928/67c6d5f4-e39b-11e5-85e6-c0378b5dd896.png)

As you can see, my Firefox put the `<img>` tag args in a different order than they were in your test, so the test failed erroneously.  HTML says you can do that, so your test has to not make assumptions about the ordering of element arguments.  A long time ago, i wrote up a wiki about writing QUnit tests - https://github.com/buttonmen-dev/buttonmen/wiki/Developer-guide:-writing-QUnit-tests - i don't know how useful it is overall, but i put some hints about DOM inspection in there.  I think the big thing is, if you expect a particular parameter to be there or not be there, do a true/false check using `match()` on the HTML output, rather than comparing full strings.

cgolubi1 commented:
Okay, here's my recommendation for an extremely minimal set of end-to-end UI checks on the vacation behavior --- let's just put responder001 on vacation, and check for the airplane in an existing page rendering test containing that player:
```diff
$ git diff test/
diff --git a/test/src/api/responderTest.php b/test/src/api/responderTest.php
index cd2f656..44fae2a 100644
--- a/test/src/api/responderTest.php
+++ b/test/src/api/responderTest.php
@@ -1284,7 +1284,7 @@ class responderTest extends PHPUnit_Framework_TestCase {
             'dob_day' => '29',
             'gender' => '',
             'comment' => '',
-            'vacation_message' => '',
+            'vacation_message' => 'player 1 is on vacation',
             'homepage' => '',
             'autoaccept' => 'true',
             'autopass' => 'false',
@@ -1305,6 +1305,7 @@ class responderTest extends PHPUnit_Framework_TestCase {
         $this->assertEquals(array_keys($retval['data']), array('playerId'));
 
         $_SESSION = $this->mock_test_user_login('responder002');
+        $args['vacation_message'] = '';
         $retval = $this->verify_api_success($args);
         $this->assertEquals($retval['status'], 'ok');
         $this->assertEquals($retval['message'], 'Player info updated successfully.');
@@ -2366,6 +2367,7 @@ class responderTest extends PHPUnit_Framework_TestCase {
 
         // Initial expected game data object
         $expData = $this->generate_init_expected_data_array($gameId, 'responder001', 'responder002', 4, 'SPECIFY_DICE');
+        $expData['playerDataArray'][0]['isOnVacation'] = TRUE;
         $expData['playerDataArray'][0]['button'] = array('name' => 'Frasquito', 'recipe' => '(4) (6) (8) (12) (2/20)', 'artFilename' => 'frasquito.png');
         $expData['playerDataArray'][1]['button'] = array('name' => 'Wiseman', 'recipe' => '(20) (20) (20) (20)', 'artFilename' => 'wiseman.png');
         $expData['playerDataArray'][0]['activeDieArray'] = array(
diff --git a/test/src/ui/js/test_Game.js b/test/src/ui/js/test_Game.js
index 599e0a5..b472d91 100644
--- a/test/src/ui/js/test_Game.js
+++ b/test/src/ui/js/test_Game.js
@@ -258,6 +258,7 @@ test("test_Game.showStatePage_newgame_nonplayer", function(assert) {
     var htmlout = Game.page.html();
     assert.ok(htmlout.length > 0,
       "The created page should have nonzero contents");
+    assert.ok(htmlout.match('vacation.png'), "The game UI shows a vacation icon when one player is on vacation");
     start();
   });
 });
```

This passes for me --- it's not a full test, but it means that if we entirely break either the ability for on-vacation players to create games, or the grabbing of the vacation info from the API and the use of that info to decide whether to draw the airplane in games, the tests might tell us.

cgolubi1 commented:
Okay, summary of all my suggestions from above:

Test suggestions:
* Fix the QUnit test `test_Overview.addPlayerCol()` so that it does not depend on browser ordering of `<img>` keyword args
* In `responderTest`, set `isOnVacation: FALSE` rather than `isOnVacation: 0`, to set the expectation for developers that this is a boolean value at the API layer.
* Add a responderTest + a UI test with a player who is on vacation, to lock in the end-to-end propagation of "is on vacation" info from `savePlayerInfo` to the UI.
* New idea: also add a UI test for a game that doesn't contain responder001, and verify that that game does *not* contain the airplane icon.  I'm pretty sure that's what would have caught the toggle regression you introduced earlier (in which everyone was on vacation after saving player info), so since we know that's a regression that's possible, let's test against it specifically.

Functionality suggestions:
* Include the vacation icon on the Open Games page
* Use the icon or color on the Profile page to draw attention to the fact that the player whose profile this is, is on vacation.
* I think we'll also eventually want something obvious that says "you are on vacation", since as the UI stands, it'd be easy to leave your vacation message on and forget that you did so.  But i'm not sure that's trivial with the API as it stands.

You don't have to do all of these things before i accept the pull (except the ones like the broken test which are bugs) --- do some of them now, and we can keep some to be later modifications, just let me know what you decide to address or not.

dwvanstone commented:
Thanks for all of the testing!

> Fix the QUnit test test_Overview.addPlayerCol() so that it does not depend on browser ordering of <img> keyword args

Fixed. (Although I can't test the fix.)

> In responderTest, set isOnVacation: FALSE rather than isOnVacation: 0, to set the expectation for developers that this is a boolean value at the API layer.

Fixed. 

> Add a responderTest + a UI test with a player who is on vacation, to lock in the end-to-end propagation of "is on vacation" info from savePlayerInfo to the UI.

This will take me a while to figure out, so I'm going to hold on this until next time.

> New idea: also add a UI test for a game that doesn't contain responder001, and verify that that game does not contain the airplane icon. I'm pretty sure that's what would have caught the toggle regression you introduced earlier (in which everyone was on vacation after saving player info), so since we know that's a regression that's possible, let's test against it specifically.

I'll hold on to this for next time as well.

> Functionality suggestions:
> Include the vacation icon on the Open Games page

Done. 

> Use the icon or color on the Profile page to draw attention to the fact that the player whose profile this is, is on vacation.

Great idea. I've also added the vacation icon to the profile name at the top to draw the connection.

> I think we'll also eventually want something obvious that says "you are on vacation", since as the UI stands, it'd be easy to leave your vacation message on and forget that you did so. But i'm not sure that's trivial with the API as it stands.

I'll hold off on this for now and try to figure this out if it becomes a priority.

A suggestion: when I run the buttonmen tests on my client (/usr/local/bin/run_buttonmen_tests) could it also suggest how to run the jshint tests? (In the same line where it says how to run QUnit test.) I usually have one or more fixes and for now I'm having to wait for a Jenkins run to find them.

Down the line we'll probably add more icons (Friend, Currently Playing Against, etc.) This fix sets the precedent of adding the icon before the name. WIll that be okay for future icons as well?

blackshadowshade commented:
To run the JSHint tests, just run grunt from util/grunt.

dwvanstone commented:
> To run the JSHint tests, just run grunt from util/grunt.

Great, that worked well!

cgolubi1 commented:
Okay, i'll look at these fixes, but i may write some end-to-end tests and give them to you to include, because i think it's a good idea to integration-test this as soon as it goes in so we don't break it later.

dwvanstone commented:
Note that the icon came from here:

https://www.iconfinder.com/icons/216191/airplane_icon#size=16

It claims to be free for commercial use, so I don't think we need to give credit to them. But if we do have an acknowledgments page sometime we can thank "Greepit.com" for the icon.

cgolubi1 commented:
Okay, so i'd like to merge this soon so devious doesn't have to do another merge from master, and so that dev testers can look at the vacation stuff for awhile during April.  Where are we with it?  It looks like:
* I need to test the latest code to see if it fixes all the problems i saw
* devious didn't want to take on adding end-to-end UI tests with the vacation icon, but i'd like to see those done because they would have caught a bug.  so the best bet is for me to come up with the UI tests, and, assuming they don't catch any more bugs, submit them as a separate pull.

Also, the latest passing jenkins is: http://jenkins.buttonweavers.com:8080/job/buttonmen-dwvanstone/54/ - try to link to jenkins when you throw code over the fence.

cgolubi1 commented on src/ui/js/Env.js:
```diff
@@ -498,6 +498,16 @@ Env.buildProfileLink = function(playerName, textOnly) {
+Env.buildVacationImage = function(size) {
+  var image = (size=='large') ? 'vacation22.png' : 'vacation16.png';
+  return  $('<img>', {
+    'src': Env.ui_root + 'images/'+image,
```
Nit (don't bother to fix unless i find actual problems): i feel strongly about horizontal whitespace around code symbols, as around the second + here.

cgolubi1 commented on src/ui/js/Profile.js:
```diff
@@ -51,10 +51,16 @@ Profile.showPage = function() {
 Profile.buildProfileTable = function() {
   var table = $('<table>', { 'class': 'profileTable', });
 
+  var profileName=$('<spane>');
```
Hmm, what is `<spane>`?  Is that an HTML element?  Once i bring the site up, i'll see if firebug recognizes it, but i suspect it's a typo for `<span>`.

cgolubi1 commented on src/ui/js/Profile.js:
```diff
@@ -122,6 +128,15 @@ Profile.buildProfileTable = function() {
+    vacationHolder = $('<span>');
+    var cookedVacation =
+      Env.prepareRawTextForDisplay(Api.profile_info.vacation_message);
+    //    vacationHolder.append(Env.buildVacationImage('small'));
```
Remove commented-out code.

cgolubi1 commented:
Test suggestions:
* Fix the QUnit test test_Overview.addPlayerCol() so that it does not depend on browser ordering of <img> keyword args
* In responderTest, set isOnVacation: FALSE rather than isOnVacation: 0, to set the expectation for developers that this is a boolean value at the API layer.
* Add a responderTest + a UI test with a player who is on vacation, to lock in the end-to-end propagation of "is on vacation" info from savePlayerInfo to the UI.
* New idea: also add a UI test for a game that doesn't contain responder001, and verify that that game does not contain the airplane icon. I'm pretty sure that's what would have caught the toggle regression you introduced earlier (in which everyone was on vacation after saving player info), so since we know that's a regression that's possible, let's test against it specifically.

Functionality suggestions:

* Include the vacation icon on the Open Games page
* Use the icon or color on the Profile page to draw attention to the fact that the player whose profile this is, is on vacation.
* I think we'll also eventually want something obvious that says "you are on vacation", since as the UI stands, it'd be easy to leave your vacation message on and forget that you did so. But i'm not sure that's trivial with the API as it stands.

You don't have to do all of these things before i accept the pull (except the ones like the broken test which are bugs) --- do some of them now, and we can keep some to be later modifications, just let me know what you decide to address or not.

cgolubi1 commented:
Test suggestions:
* ~~Fix the QUnit test test_Overview.addPlayerCol() so that it does not depend on browser ordering of <img> keyword args~~: looks reasonable, and passes under QUnit for me now.
* ~~In responderTest, set isOnVacation: FALSE rather than isOnVacation: 0, to set the expectation for developers that this is a boolean value at the API layer~~: this was done
* Add a responderTest + a UI test with a player who is on vacation, to lock in the end-to-end propagation of "is on vacation" info from savePlayerInfo to the UI: not done, i'll work on this
* New idea: also add a UI test for a game that doesn't contain responder001, and verify that that game does not contain the airplane icon. I'm pretty sure that's what would have caught the toggle regression you introduced earlier (in which everyone was on vacation after saving player info), so since we know that's a regression that's possible, let's test against it specifically: not done, i'll work on this

Functionality suggestions:

* ~~Include the vacation icon on the Open Games page~~: this was done, and seems to work
* ~~Use the icon or color on the Profile page to draw attention to the fact that the player whose profile this is, is on vacation~~: this was done
* I think we'll also eventually want something obvious that says "you are on vacation", since as the UI stands, it'd be easy to leave your vacation message on and forget that you did so. But i'm not sure that's trivial with the API as it stands: not done, and we're not worrying about it right now - let's see what testers think

cgolubi1 commented:
Okay, so i had a couple of code nits above, and i *do* think `<spane>` needs to be fixed unless someone tells me that's a real thing.  Google didn't obviously believe it was a real thing, and i would worry that there are browsers out there that might do something unexpected with an unrecognized element tag.

cgolubi1 commented:
Okay, devious, here's a git diff for the minimal end-to-end regression test i had in mind --- basically it's four changes:
* in `responderTest.php` function `test_request_savePlayerInfo()`, save a vacation message for player responder002
* in `responderTest.php` function `test_api_game_001()` (the only game containing responder002), make sure the `isOnVacation` flag not set for responder002
* in `test_Game.js` function `test_Game.showStatePage()` (which uses API data for a game containing responder002), make sure the vacation icon is visible
* in `test_Game.js` function `test_Game.showStatePage_swingset()` (which uses API data for a game not containing responder002), make sure the vacation icon not is visible

If that seems reasonable to you, want to integrate those changes into your pull when you fix the `<spane>` typo?  

```diff
diff --git a/test/src/api/responderTest.php b/test/src/api/responderTest.php
index fe349d3..069321d 100644
--- a/test/src/api/responderTest.php
+++ b/test/src/api/responderTest.php
@@ -1305,7 +1305,9 @@ class responderTest extends PHPUnit_Framework_TestCase {
         $this->assertEquals(array_keys($retval['data']), array('playerId'));
 
         $_SESSION = $this->mock_test_user_login('responder002');
+        $args['vacation_message'] = 'Player 2 is on vacation';
         $retval = $this->verify_api_success($args);
+        $args['vacation_message'] = '';
         $this->assertEquals($retval['status'], 'ok');
         $this->assertEquals($retval['message'], 'Player info updated successfully.');
         $this->assertEquals(array_keys($retval['data']), array('playerId'));
@@ -2384,6 +2386,9 @@ class responderTest extends PHPUnit_Framework_TestCase {
         $expData['playerDataArray'][1]['waitingOnAction'] = FALSE;
         $expData['playerDataArray'][0]['optRequestArray'] = array('4' => array(2, 20));
 
+        // in test_request_savePlayerInfo() responder002 was set to be on vacation - make sure the game reflects that.
+        $expData['playerDataArray'][1]['isOnVacation'] = TRUE;
+
         // now load the game and check its state
         $retval = $this->verify_api_loadGameData($expData, $gameId, 10);
 
diff --git a/test/src/ui/js/test_Game.js b/test/src/ui/js/test_Game.js
index 599e0a5..37e0021 100644
--- a/test/src/ui/js/test_Game.js
+++ b/test/src/ui/js/test_Game.js
@@ -163,6 +163,8 @@ test("test_Game.showStatePage", function(assert) {
     var htmlout = Game.page.html();
     assert.ok(htmlout.length > 0,
       "The created page should have nonzero contents");
+    assert.ok(htmlout.match('vacation16.png'),
+      "The game UI contains a vacation icon when the API data reports that one player is on vacation");
     start();
   });
 });
@@ -246,6 +248,8 @@ test("test_Game.showStatePage_swingset", function(assert) {
     var htmlout = Game.page.html();
     assert.ok(htmlout.length > 0,
       "The created page should have nonzero contents");
+    assert.ok(!htmlout.match('vacation16.png'),
+      "The game UI does not contain a vacation icon when the API data reports that neither player is on vacation");
     start();
   });
 });

cgolubi1 commented:
I think that's it, and this will be ready to merge once it comes back from devious.

dwvanstone commented:
Changes made, Jenkins passed!

http://jenkins.buttonweavers.com:8080/job/buttonmen-dwvanstone/55/

cgolubi1 commented:
Yay!  Thanks for working so hard on this.  Merging now, and let's see what the dev testers think!
